File: Core/Models/Operations.swift
import Foundation

public enum AnnouncementAudience: String, Codable, CaseIterable, Sendable, Hashable {
    case all
    case coaches
    case parents
    case athletes
    case branchManagers

    public var labelKey: String { "announcement.audience.\(rawValue)" }
}

public struct Announcement: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var branchID: EntityID?
    public var title: String
    public var titleAr: String
    public var body: String
    public var bodyAr: String
    public var audience: AnnouncementAudience
    public var publishedAt: Date
    public var publishedByUserID: EntityID
    public var requiresRSVP: Bool
    public var rsvpDeadline: Date?

    // MARK: Stage 1.9 — announcements dashboard
    /// Lifecycle state — drives the status pill and the top filter pills.
    public var status: AnnouncementStatus
    /// Editorial category — drives the row's pastel icon tile.
    public var category: AnnouncementCategory
    /// Asset-catalogue hero image; `nil` falls back to a category gradient.
    public var imageAssetName: String?
    /// When a scheduled announcement is set to broadcast.
    public var scheduledAt: Date?
    /// Full audience list shown as chips. Empty → derive from `audience`.
    public var audiences: [AnnouncementAudience]
    public var location: String?
    public var eventStart: Date?
    public var eventEnd: Date?
    public var registrationDeadline: Date?
    /// Per-channel delivery outcomes.
    public var delivery: [AnnouncementDelivery]
    /// Reach + interaction counters for the Engagement section.
    public var engagement: AnnouncementEngagement?
    public var attachments: [AnnouncementAttachment]
    /// Display name of the author ("Dr Ali Alawi").
    public var authorName: String?

    public init(
        id: EntityID = UUID(),
        branchID: EntityID? = nil,
        title: String,
        titleAr: String,
        body: String,
        bodyAr: String,
        audience: AnnouncementAudience,
        publishedAt: Date = Date(),
        publishedByUserID: EntityID,
        requiresRSVP: Bool = false,
        rsvpDeadline: Date? = nil,
        status: AnnouncementStatus = .published,
        category: AnnouncementCategory = .general,
        imageAssetName: String? = nil,
        scheduledAt: Date? = nil,
        audiences: [AnnouncementAudience] = [],
        location: String? = nil,
        eventStart: Date? = nil,
        eventEnd: Date? = nil,
        registrationDeadline: Date? = nil,
        delivery: [AnnouncementDelivery] = [],
        engagement: AnnouncementEngagement? = nil,
        attachments: [AnnouncementAttachment] = [],
        authorName: String? = nil
    ) {
        self.id = id
        self.branchID = branchID
        self.title = title
        self.titleAr = titleAr
        self.body = body
        self.bodyAr = bodyAr
        self.audience = audience
        self.publishedAt = publishedAt
        self.publishedByUserID = publishedByUserID
        self.requiresRSVP = requiresRSVP
        self.rsvpDeadline = rsvpDeadline
        self.status = status
        self.category = category
        self.imageAssetName = imageAssetName
        self.scheduledAt = scheduledAt
        self.audiences = audiences
        self.location = location
        self.eventStart = eventStart
        self.eventEnd = eventEnd
        self.registrationDeadline = registrationDeadline
        self.delivery = delivery
        self.engagement = engagement
        self.attachments = attachments
        self.authorName = authorName
    }

    /// Backward-compatible decoder — pre-1.9 rows (no dossier columns) parse
    /// with sensible defaults.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        self.id = try c.decode(EntityID.self, forKey: .id)
        self.branchID = try c.decodeIfPresent(EntityID.self, forKey: .branchID)
        self.title = try c.decode(String.self, forKey: .title)
        self.titleAr = try c.decodeIfPresent(String.self, forKey: .titleAr) ?? ""
        self.body = try c.decodeIfPresent(String.self, forKey: .body) ?? ""
        self.bodyAr = try c.decodeIfPresent(String.self, forKey: .bodyAr) ?? ""
        self.audience = try c.decode(AnnouncementAudience.self, forKey: .audience)
        self.publishedAt = try c.decode(Date.self, forKey: .publishedAt)
        self.publishedByUserID = try c.decode(EntityID.self, forKey: .publishedByUserID)
        self.requiresRSVP = try c.decodeIfPresent(Bool.self, forKey: .requiresRSVP) ?? false
        self.rsvpDeadline = try c.decodeIfPresent(Date.self, forKey: .rsvpDeadline)
        self.status = try c.decodeIfPresent(AnnouncementStatus.self, forKey: .status) ?? .published
        self.category = try c.decodeIfPresent(AnnouncementCategory.self, forKey: .category) ?? .general
        self.imageAssetName = try c.decodeIfPresent(String.self, forKey: .imageAssetName)
        self.scheduledAt = try c.decodeIfPresent(Date.self, forKey: .scheduledAt)
        self.audiences = try c.decodeIfPresent([AnnouncementAudience].self, forKey: .audiences) ?? []
        self.location = try c.decodeIfPresent(String.self, forKey: .location)
        self.eventStart = try c.decodeIfPresent(Date.self, forKey: .eventStart)
        self.eventEnd = try c.decodeIfPresent(Date.self, forKey: .eventEnd)
        self.registrationDeadline = try c.decodeIfPresent(Date.self, forKey: .registrationDeadline)
        self.delivery = try c.decodeIfPresent([AnnouncementDelivery].self, forKey: .delivery) ?? []
        self.engagement = try c.decodeIfPresent(AnnouncementEngagement.self, forKey: .engagement)
        self.attachments = try c.decodeIfPresent([AnnouncementAttachment].self, forKey: .attachments) ?? []
        self.authorName = try c.decodeIfPresent(String.self, forKey: .authorName)
    }

    /// Full audience list — falls back to the single `audience` for old rows.
    public var effectiveAudiences: [AnnouncementAudience] {
        audiences.isEmpty ? [audience] : audiences
    }

    /// The date the row and detail panel display.
    public var displayDate: Date {
        if status == .scheduled, let scheduledAt { return scheduledAt }
        return publishedAt
    }
}

public enum RSVPResponse: String, Codable, CaseIterable, Sendable, Hashable {
    case yes, no, maybe

    public var labelKey: String { "announcement.rsvp.\(rawValue)" }
}

public struct AnnouncementRSVP: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var announcementID: EntityID
    public var userID: EntityID
    public var response: RSVPResponse
    public var respondedAt: Date

    public init(
        id: EntityID = UUID(),
        announcementID: EntityID,
        userID: EntityID,
        response: RSVPResponse,
        respondedAt: Date = Date()
    ) {
        self.id = id
        self.announcementID = announcementID
        self.userID = userID
        self.response = response
        self.respondedAt = respondedAt
    }
}

public enum CertificationKind: String, Codable, CaseIterable, Sendable, Hashable {
    case firstAid, safeguarding, wtCoaching, doping, refereeing

    public var labelKey: String { "cert.\(rawValue)" }
}

public enum CertificationSeverity: String, Sendable, Hashable {
    case ok, expiring, expired

    public var labelKey: String { "cert.severity.\(rawValue)" }
}

public struct Certification: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var coachID: EntityID
    public var kind: CertificationKind
    public var issuer: String
    public var issuedAt: Date
    public var expiresAt: Date
    public var fileRef: String?

    public init(
        id: EntityID = UUID(),
        coachID: EntityID,
        kind: CertificationKind,
        issuer: String,
        issuedAt: Date,
        expiresAt: Date,
        fileRef: String? = nil
    ) {
        self.id = id
        self.coachID = coachID
        self.kind = kind
        self.issuer = issuer
        self.issuedAt = issuedAt
        self.expiresAt = expiresAt
        self.fileRef = fileRef
    }

    public var daysUntilExpiry: Int {
        Calendar.current.dateComponents([.day], from: Date(), to: expiresAt).day ?? 0
    }

    public var severity: CertificationSeverity {
        let days = daysUntilExpiry
        if days < 0 { return .expired }
        if days < 60 { return .expiring }
        return .ok
    }
}

File: flutter_app/lib/core/models/operations.dart
// Port of Core/Models/Operations.swift.
// Pure data — no Flutter imports (logic-layer rule).

import 'announcement_dossier.dart';
import 'entity_id.dart';

export 'announcement_dossier.dart';

enum AnnouncementAudience {
  all,
  coaches,
  parents,
  athletes,
  branchManagers;

  String get labelKey => 'announcement.audience.$name';

  static AnnouncementAudience fromJson(String raw) => AnnouncementAudience
      .values
      .firstWhere((e) => e.name == raw, orElse: () => AnnouncementAudience.all);
}

class Announcement {
  final EntityID id;
  final EntityID? branchId;
  final String title;
  final String titleAr;
  final String body;
  final String bodyAr;
  final AnnouncementAudience audience;
  final DateTime publishedAt;
  final EntityID publishedByUserId;
  final bool requiresRsvp;
  final DateTime? rsvpDeadline;

  // Stage 1.9 — announcements dashboard
  final AnnouncementStatus status;
  final AnnouncementCategory category;
  final String? imageAssetName;
  final DateTime? scheduledAt;
  final List<AnnouncementAudience> audiences;
  final String? location;
  final DateTime? eventStart;
  final DateTime? eventEnd;
  final DateTime? registrationDeadline;
  final List<AnnouncementDelivery> delivery;
  final AnnouncementEngagement? engagement;
  final List<AnnouncementAttachment> attachments;
  final String? authorName;

  const Announcement({
    required this.id,
    this.branchId,
    required this.title,
    this.titleAr = '',
    this.body = '',
    this.bodyAr = '',
    required this.audience,
    required this.publishedAt,
    required this.publishedByUserId,
    this.requiresRsvp = false,
    this.rsvpDeadline,
    this.status = AnnouncementStatus.published,
    this.category = AnnouncementCategory.general,
    this.imageAssetName,
    this.scheduledAt,
    this.audiences = const [],
    this.location,
    this.eventStart,
    this.eventEnd,
    this.registrationDeadline,
    this.delivery = const [],
    this.engagement,
    this.attachments = const [],
    this.authorName,
  });

  /// Full audience list — falls back to the single [audience] for old rows.
  List<AnnouncementAudience> get effectiveAudiences =>
      audiences.isEmpty ? [audience] : audiences;

  /// The date the row and detail panel display.
  DateTime get displayDate {
    if (status == AnnouncementStatus.scheduled && scheduledAt != null) {
      return scheduledAt!;
    }
    return publishedAt;
  }

  factory Announcement.fromJson(Map<String, dynamic> json) => Announcement(
    id: json['id'] as String,
    branchId: json['branchID'] as String?,
    title: json['title'] as String,
    titleAr: json['titleAr'] as String? ?? '',
    body: json['body'] as String? ?? '',
    bodyAr: json['bodyAr'] as String? ?? '',
    audience: AnnouncementAudience.fromJson(
      json['audience'] as String? ?? 'all',
    ),
    publishedAt: DateTime.parse(json['publishedAt'] as String),
    publishedByUserId: json['publishedByUserID'] as String,
    requiresRsvp: json['requiresRSVP'] as bool? ?? false,
    rsvpDeadline: json['rsvpDeadline'] != null
        ? DateTime.parse(json['rsvpDeadline'] as String)
        : null,
    status: AnnouncementStatus.fromJson(
      json['status'] as String? ?? 'published',
    ),
    category: AnnouncementCategory.fromJson(
      json['category'] as String? ?? 'general',
    ),
    imageAssetName: json['imageAssetName'] as String?,
    scheduledAt: json['scheduledAt'] != null
        ? DateTime.parse(json['scheduledAt'] as String)
        : null,
    audiences: ((json['audiences'] as List?) ?? [])
        .map((e) => AnnouncementAudience.fromJson(e as String))
        .toList(),
    location: json['location'] as String?,
    eventStart: json['eventStart'] != null
        ? DateTime.parse(json['eventStart'] as String)
        : null,
    eventEnd: json['eventEnd'] != null
        ? DateTime.parse(json['eventEnd'] as String)
        : null,
    registrationDeadline: json['registrationDeadline'] != null
        ? DateTime.parse(json['registrationDeadline'] as String)
        : null,
    delivery: ((json['delivery'] as List?) ?? [])
        .map((e) => AnnouncementDelivery.fromJson(e as Map<String, dynamic>))
        .toList(),
    engagement: json['engagement'] != null
        ? AnnouncementEngagement.fromJson(
            json['engagement'] as Map<String, dynamic>,
          )
        : null,
    attachments: ((json['attachments'] as List?) ?? [])
        .map((e) => AnnouncementAttachment.fromJson(e as Map<String, dynamic>))
        .toList(),
    authorName: json['authorName'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'branchID': branchId,
    'title': title,
    'titleAr': titleAr,
    'body': body,
    'bodyAr': bodyAr,
    'audience': audience.name,
    'publishedAt': publishedAt.toIso8601String(),
    'publishedByUserID': publishedByUserId,
    'requiresRSVP': requiresRsvp,
    'rsvpDeadline': rsvpDeadline?.toIso8601String(),
    'status': status.name,
    'category': category.name,
    'imageAssetName': imageAssetName,
    'scheduledAt': scheduledAt?.toIso8601String(),
    'audiences': audiences.map((a) => a.name).toList(),
    'location': location,
    'eventStart': eventStart?.toIso8601String(),
    'eventEnd': eventEnd?.toIso8601String(),
    'registrationDeadline': registrationDeadline?.toIso8601String(),
    'delivery': delivery.map((d) => d.toJson()).toList(),
    'engagement': engagement?.toJson(),
    'attachments': attachments.map((a) => a.toJson()).toList(),
    'authorName': authorName,
  };
}

enum RSVPResponse {
  yes,
  no,
  maybe;

  String get labelKey => 'announcement.rsvp.$name';

  static RSVPResponse fromJson(String raw) => RSVPResponse.values.firstWhere(
    (e) => e.name == raw,
    orElse: () => RSVPResponse.maybe,
  );
}

class AnnouncementRSVP {
  final EntityID id;
  final EntityID announcementId;
  final EntityID userId;
  final RSVPResponse response;
  final DateTime respondedAt;

  const AnnouncementRSVP({
    required this.id,
    required this.announcementId,
    required this.userId,
    required this.response,
    required this.respondedAt,
  });

  factory AnnouncementRSVP.fromJson(Map<String, dynamic> json) =>
      AnnouncementRSVP(
        id: json['id'] as String,
        announcementId: json['announcementID'] as String,
        userId: json['userID'] as String,
        response: RSVPResponse.fromJson(json['response'] as String),
        respondedAt: DateTime.parse(json['respondedAt'] as String),
      );

  Map<String, dynamic> toJson() => {
    'id': id,
    'announcementID': announcementId,
    'userID': userId,
    'response': response.name,
    'respondedAt': respondedAt.toIso8601String(),
  };
}

enum CertificationKind {
  firstAid,
  safeguarding,
  wtCoaching,
  doping,
  refereeing;

  String get labelKey => 'cert.$name';

  /// Grouping shown under the branch in the certifications table.
  String get categoryLabelKey => 'cert.category.$name';

  /// SF Symbol for the row's status icon tile.
  String get systemIcon => switch (this) {
    CertificationKind.firstAid => 'cross.case.fill',
    CertificationKind.safeguarding => 'shield.lefthalf.filled',
    CertificationKind.wtCoaching => 'rosette',
    CertificationKind.doping => 'drop.fill',
    CertificationKind.refereeing => 'flag.checkered',
  };

  static CertificationKind fromJson(String raw) =>
      CertificationKind.values.firstWhere(
        (e) => e.name == raw,
        orElse: () => CertificationKind.wtCoaching,
      );
}

/// Not Codable in Swift — plain enum, no fromJson/toJson.
enum CertificationSeverity {
  ok,
  expiring,
  expired;

  String get labelKey => 'cert.severity.$name';
}

class Certification {
  final EntityID id;
  final EntityID coachId;
  final CertificationKind kind;
  final String issuer;
  final DateTime issuedAt;
  final DateTime expiresAt;
  final String? fileRef;

  const Certification({
    required this.id,
    required this.coachId,
    required this.kind,
    required this.issuer,
    required this.issuedAt,
    required this.expiresAt,
    this.fileRef,
  });

  int get daysUntilExpiry => expiresAt.difference(DateTime.now()).inDays;

  CertificationSeverity get severity {
    final days = daysUntilExpiry;
    if (days < 0) return CertificationSeverity.expired;
    if (days < 60) return CertificationSeverity.expiring;
    return CertificationSeverity.ok;
  }

  factory Certification.fromJson(Map<String, dynamic> json) => Certification(
    id: json['id'] as String,
    coachId: json['coachID'] as String,
    kind: CertificationKind.fromJson(json['kind'] as String),
    issuer: json['issuer'] as String,
    issuedAt: DateTime.parse(json['issuedAt'] as String),
    expiresAt: DateTime.parse(json['expiresAt'] as String),
    fileRef: json['fileRef'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'coachID': coachId,
    'kind': kind.name,
    'issuer': issuer,
    'issuedAt': issuedAt.toIso8601String(),
    'expiresAt': expiresAt.toIso8601String(),
    'fileRef': fileRef,
  };
}
